lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,506 @@
+		if totalTxnsSize > maxSizeOfNumBlocks {
+			break
+		}
+	}
+	globalMinFeeRate, _ := txnRegister.minimumNetworkFeeNanosPerKB.Uint64()
+	if len(txns) == 0 {
+		// If there are no txns in the transaction register, we simply return the minimum network fee.
+		return globalMinFeeRate, nil
+	}
+
+	bucketMinFee, bucketMaxFee, err := getPriorityFeeBucketFromTxns(
+		txns,
+		priorityPercentileBasisPoints,
+		txnRegister.minimumNetworkFeeNanosPerKB,
+		txnRegister.feeBucketGrowthRateBasisPoints)
+	if err != nil {
+		return 0, errors.Wrap(err,
+			"estimateFeeRateNanosPerKBGivenTransactionRegister: Problem computing priority fee bucket")
+	}
+	// If the bucketMinFee is less than or equal to the global min fee rate, we return the global min fee rate.
+	if bucketMinFee <= globalMinFeeRate {
+		return globalMinFeeRate, nil
+	}
+
+	// Compute the congestion threshold. If our congestion factor is 100% (or 10,000 bps),
+	// then congestion threshold is simply max block size * numBlocks
+	// TODO: I don't know if I like this name really.
+	congestionThreshold := (congestionFactorBasisPoints * maxSizeOfNumBlocks) / MaxBasisPoints
+	// If the total size of the txns in the transaction register is less than the computed congestion threshold,
+	// we return one bucket lower than the Priority fee.
+	if totalTxnsSize <= congestionThreshold {
+		// Return one bucket lower than Priority fee
+		oneBucketLowerMinFee, _ := computeFeeTimeBucketRangeFromFeeNanosPerKB(
+			bucketMinFee-1,
+			txnRegister.minimumNetworkFeeNanosPerKB,
+			txnRegister.feeBucketGrowthRateBasisPoints,
+		)
+		return oneBucketLowerMinFee, nil
+	}
+	// If the total size of the txns in the mempool is greater than the computed congestion threshold
+	// but less than the max size of num blocks, we return the Priority fee.
+	if totalTxnsSize > congestionThreshold && totalTxnsSize <= maxSizeOfNumBlocks {
+		// Return Priority fee
+		return bucketMinFee, nil
+	}
+	// Otherwise, we return one bucket higher than Priority fee
+	return bucketMaxFee + 1, nil
+}
+
+// getPriorityFeeBucketFromTxns computes the priority fee bucket for the given transactions using the
+// priorityPercentileBasisPoints, minimumNetworkFeeNanosPerKB, and feeBucketGrowthRateBasisPoints params.
+// The feeTimeOrderedTxns have the highest fees first and the lowest fees last, so we need to compute
+// the percentile position of the priorityPercentileBasisPoints param and then compute the fee bucket
+// range based on the fee rate per KB of the transaction at that position.
+func getPriorityFeeBucketFromTxns(
+	feeTimeOrderedTxns []*MempoolTx,
+	priorityPercentileBasisPoints uint64,
+	minimumNetworkFeeNanosPerKB *big.Float,
+	feeBucketGrowthRateBasisPoints *big.Float,
+) (uint64, uint64, error) {
+	percentilePosition := uint64(
+		len(feeTimeOrderedTxns)) - ((priorityPercentileBasisPoints * uint64(len(feeTimeOrderedTxns))) / 10000)
+	if percentilePosition >= uint64(len(feeTimeOrderedTxns)) {
```

Comment:
This will happen if the priority is zero. In that case... probably what they want is `len(feeTimeOrderedTxns)-1`? In which case consider just returning that instead of erroring here.